content/en/blog/_posts/2020-03-30-topology-manager-beta.md

```diff
@@ -0,0 +1,502 @@
+
+```
+{cpu: {0, 1}, gpu: 0, nic: 0}
+{cpu: {0, 2}, gpu: 0, nic: 0}
+{cpu: {0, 3}, gpu: 0, nic: 0}
+{cpu: {1, 2}, gpu: 0, nic: 0}
+{cpu: {1, 3}, gpu: 0, nic: 0}
+{cpu: {2, 3}, gpu: 0, nic: 0}
+
+{cpu: {4, 5}, gpu: 1, nic: 1}
+{cpu: {4, 6}, gpu: 1, nic: 1}
+{cpu: {4, 7}, gpu: 1, nic: 1}
+{cpu: {5, 6}, gpu: 1, nic: 1}
+{cpu: {5, 7}, gpu: 1, nic: 1}
+{cpu: {6, 7}, gpu: 1, nic: 1}
+```
+
+And that’s it! Just follow this pattern to have the **<code>TopologyManager</code>** ensure NUMA alignment across containers that request topology-aware devices and exclusive CPUs.
+
+**NOTE:** if a pod is rejected by one of the **<code>TopologyManager</code>** policies, it will be placed in a **<code>Terminated</code>** state with a pod admission error and a reason of "**<code>TopologyAffinityError</code>**". Once a pod is in this state, the Kubernetes scheduler will not attempt to reschedule it. It is therefore recommended to use a [**<code>Deployment</code>**](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#creating-a-deployment) with replicas to trigger a redeploy of the pod on such a failure. An [external control loop](https://kubernetes.io/docs/concepts/architecture/controller/) can also be implemented to trigger a redeployment of pods that have a **<code>TopologyAffinityError</code>**.
+
+## This is great, so how does it work under the hood?
+
+Pseudocode for the primary logic carried out by the **<code>TopologyManager</code>** can be seen below:
+
+```
+for container := range append(InitContainers, Containers...) {
+    for provider := range HintProviders {
+        hints += provider.GetTopologyHints(container)
+    }
+
+    bestHint := policy.Merge(hints)
+
+    for provider := range HintProviders {
+        provider.Allocate(container, bestHint)
+    }
+}
+```
+
+The following diagram summarizes the steps taken during this loop:
+
+<p align="center">
+  <img weight="200" height="200" src="/images/blog/2020-03-25-kubernetes-1.18-release-announcement/numa-steps-during-loop.png">
+</p>
+
+The steps themselves are:
+1. Loop over all containers in a pod.
+1. For each container, gather "**<code>TopologyHints</code>**" from a set of "**<code>HintProviders</code>**" for each topology-aware resource type requested by the container (e.g. **<code>gpu-vendor.com/gpu</code>**, **<code>nic-vendor.com/nic</code>**, **<code>cpu</code>**, etc.).
+1. Using the selected policy, merge the gathered **<code>TopologyHints</code>** to find the "best" hint that aligns resource allocations across all resource types.
+1. Loop back over the set of hint providers, instructing them to allocate the resources they control using the merged hint as a guide.
+1. This loop runs at pod admission time and will fail to admit the pod if any of these steps fail or alignment cannot be satisfied according to the selected policy. Any resources allocated before the failure are cleaned up accordingly.
+
```

Comment:
Looks like this is not rendering either as bullets or as numbered list. Probably just needs a line break after `The steps themselves are:`